File: 0399-evaluate-division/0399-evaluate-division.py
class Solution:
    def calcEquation(self, equations, values, queries):
        visited = set()

        def dfs(curr, dest):
            visited.add(curr)
            if curr == dest: return 1
            ret_val = -1
            for d in graph[curr]:
                if dest == d[0]:
                    # print(curr, "->", d[0], "(", d[1], ")", end="\t")
                    # print("END")
                    return d[1]
                elif d[0] not in visited:
                    # print(curr, "->", d[0], "(", d[1], ")", end="\t")
                    test = dfs(d[0], dest)
                    if test != -1:
                        ret_val = d[1] * test
            # print("END")
            return ret_val

        graph = defaultdict(list)
        for i, eq in enumerate(equations):
            graph[eq[0]] += [(eq[1], values[i])]
            graph[eq[1]] += [(eq[0], 1 / values[i])]
        res = []
        # import pprint; pprint.pprint(graph)
        for q in queries:
            if q[0] not in graph or q[1] not in graph:
                res.append(-1)
            else:
                # print(q)
                res.append(dfs(q[0], q[1]))
                visited.clear()
                # print("\n" * 3)
        return res

File: 0399-evaluate-division/0399-evaluate-division.java
import java.util.*;

class Solution {
    Map<String, List<Pair>> graph;

    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        graph = new HashMap<>();
        for (int i = 0; i < equations.size(); i++) {
            List<String> eq = equations.get(i);
            String src = eq.get(0);
            String dest = eq.get(1);
            double value = values[i];
            graph.putIfAbsent(src, new ArrayList<>());
            graph.putIfAbsent(dest, new ArrayList<>());
            graph.get(src).add(new Pair(dest, value));
            graph.get(dest).add(new Pair(src, 1.0 / value));
        }

        double[] result = new double[queries.size()];
        for (int i = 0; i < queries.size(); i++) {
            List<String> query = queries.get(i);
            String src = query.get(0);
            String dest = query.get(1);
            if (!graph.containsKey(src) || !graph.containsKey(dest)) {
                result[i] = -1.0;
            } else {
                result[i] = dfs(src, dest, new HashSet<>());
            }
        }
        return result;
    }

    private double dfs(String curr, String dest, Set<String> visited) {
        visited.add(curr);
        if (curr.equals(dest)) {
            return 1.0;
        }
        double ret_val = -1.0;
        for (Pair neighbor : graph.get(curr)) {
            if (!visited.contains(neighbor.node)) {
                double test = dfs(neighbor.node, dest, visited);
                if (test != -1.0) {
                    ret_val = neighbor.value * test;
                    break;
                }
            }
        }
        visited.remove(curr);
        return ret_val;
    }

    class Pair {
        String node;
        double value;

        Pair(String node, double value) {
            this.node = node;
            this.value = value;
        }
    }
}
